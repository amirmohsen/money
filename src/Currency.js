import CurrencyStore from './CurrencyStore';
import Formatter from './Formatter';
import WrongInputError from './errors/WrongInputError';

/**
 * @example <caption>Using code</caption>
 * let currency = new Currency('USD');
 *
 * @example <caption>Using existing currency object</caption>
 * let currency = new Currency(new Currency('USD'));
 *
 * @example <caption>Using custom settings to create a one-off currency without registering it</caption>
 * let currency = new Currency({
 *  code: 'ETH',
 *  symbol: 'Îž'
 * });
 */
export default class Currency {

	/**
	 * @param {string|object|Currency} currency - Currency string code, custom settings or instance of Currency
	 */
	constructor(currency) {
		/**
		 * @type {object} - an object holding currency details such as decimal digits, etc.
		 * @private
		 */
		this._currencySettings = this._preProcess(currency);
	}

	/**
	 * Check if the parameter currency the same as the current currency
	 * @param {string|Currency} currency - Currency string code or instance of Currency
	 * @returns {boolean} - returns true if the parameter currency is the same as the current currency
	 */
	is(currency) {
		currency = new Currency(currency);
		return this.getCode() === currency.getCode();
	}

	/**
	 * Get currency settings
	 * @returns {Object} - Currency settings
	 */
	getSettings() {
		return this._currencySettings;
	}

	/**
	 * Get currency code
	 * @returns {string} - Currency code
	 */
	toString() {
		return this.getCode();
	}

	/**
	 * Get currency code
	 * @returns {string} - Currency code
	 */
	toJSON() {
		return this.getCode();
	}

	/**
	 * Clone the currency instance
	 * @returns {Currency} - new Currency instance
	 */
	clone() {
		return new this.constructor(this.getCode());
	}

	/**
	 * Get currency code
	 * @returns {string} - Currency code
	 */
	getCode() {
		return this._currencySettings.code;
	}

	/**
	 * Get currency symbol
	 * @returns {string} - Currency symbol
	 */
	getSymbol() {
		return this._currencySettings.symbol;
	}

	/**
	 * Get currency thousands separator
	 * @returns {string} - Currency thousands separator
	 */
	getThousandsSeparator() {
		return this._currencySettings.thousandsSeparator;
	}

	/**
	 * Get currency decimal separator
	 * @returns {string} - Currency decimal separator
	 */
	getDecimalSeparator() {
		return this._currencySettings.decimalSeparator;
	}

	/**
	 * Get currency formatting pattern
	 * @returns {string} - Currency format pattern
	 */
	getPattern() {
		return this._currencySettings.pattern;
	}

	/**
	 * Get the number of decimal digits for this currency
	 * @returns {number} - Number of decimal digits
	 */
	getDecimalDigits() {
		return this._currencySettings.decimalDigits;
	}

	/**
	 * Get the formatter for this currency
	 * @returns {function|null|undefined} - Currency formatter
	 */
	getFormatter() {
		return this._currencySettings.formatter;
	}

	/**
	 * Get the parser for this currency
	 * @returns {function|null|undefined} - Currency parser
	 */
	getParser() {
		return this._currencySettings.parser;
	}

	/**
	 * Format a monetary value
	 * @param {Money} value - Monetary value to be formatted
	 * @param {object} [overrideSettings] - settings to override Currency's default formatting settings
	 * @returns {string} - Formatted string of the value
	 */
	format(value, overrideSettings = {}) {
		return Formatter.format(value, new Currency({
			...this._currencySettings,
			...overrideSettings
		}));
	}

	/**
	 * Parse a monetary value from a formatted value (same as `unformat`)
	 * @param {string} value - Formatted value to be parsed into a monetary value
	 * @param {object} [overrideSettings] - settings to override Currency's default formatting settings
	 * @returns {Money} - Parsed "Money" value
	 */
	parse(value, overrideSettings = {}) {
		return Formatter.parse(value, new Currency({
			...this._currencySettings,
			...overrideSettings
		}));
	}

	/**
	 * Used by the constructor to pre-process the input
	 * @param {string|object|Currency} currency - Currency code, settings or instance of Currency
	 * @returns {object} - Currency settings
	 */
	_preProcess(currency) {
		let settings = {};

		if(currency instanceof Currency) {
			settings = CurrencyStore.get(currency.getCode());
		}
		else if(typeof currency === 'string') {
			settings = CurrencyStore.get(currency);
		}
		else if(typeof currency === 'object') {
			let defaultSettings = {};

			if(currency.code) {
				defaultSettings = CurrencyStore.get(currency.code);
			}

			settings = {
				...defaultSettings,
				...settings
			};
		}
		else {
			throw new WrongInputError('Invalid currency provided.');
		}

		return settings;
	}

	/**
	 * Register (or replace) a currency
	 * @param {string} code - Currency code
	 * @param {object} [settings] - Currency settings
	 * @param {string} [settings.symbol=code] - Currency symbol (by default same as code)
	 * @param {string} [settings.thousandsSeparator=','] - Currency thousands separator
	 * @param {string} [settings.decimalSeparator='.'] -Currency decimal separator
	 * @param {number} [settings.decimalDigits=2] - Currency decimal digits
	 * @param {string} [settings.pattern='%ns%s%v'] - Currency formatting pattern. %ns is number sign's placeholder, %s is symbol's placeholder and $v is monetary value's placeholder.
	 * @param {?function(value: Money): string} [settings.formatter] - Currency formatter, optional (used instead of the pattern)
	 * @param {?function(value: string): Money} [settings.parser] - Currency parser, optional
	 */
	static register(code, settings) {
		return CurrencyStore.set(code, settings);
	}

	/**
	 * Get currency settings for the specified currency
	 * @param {string} code - Currency code
	 * @return {object} - Currency settings
	 */
	static getSettings(code) {
		return CurrencyStore.get(code);
	}

	/**
	 * Get an array of currency settings for all currencies
	 * @return {object[]} - Settings for all currencies
	 */
	static getAllSettings() {
		return CurrencyStore.getAll();
	}
}
