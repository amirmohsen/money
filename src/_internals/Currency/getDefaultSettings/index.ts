import { CurrencySettings } from '..';

const getDefaultSettings = (code: string): CurrencySettings => ({
  code,
  symbol: code,
  thousandsSeparator: ',',
  decimalSeparator: '.',
  decimalDigits: 2,
  pattern: '%ns%s%v',
});

export default getDefaultSettings;
