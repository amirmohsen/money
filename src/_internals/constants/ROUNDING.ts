import BigNumber from 'bignumber.js';

/**
 * Rounding modes you can use in your operations.
 * These map directly to
 * [`BigNumber`'s rounding modes](http://mikemcl.github.io/bignumber.js/#constructor-properties).
 *
 * @example
 * Rounding mode usage
 * ```js
 * const
 *      price = new Money('7856', 'USD'), // $78.56
 *      discountedAndRoundedUp = price.multiply('0.70'), // $55.00
 *      discountedAndRoundedDown = price.multiply('0.70', ROUNDING.DOWN); // $54.99
 * ```
 */
enum ROUNDING {
  /**
   * Rounds away from zero
   */
  UP = BigNumber.ROUND_UP,
  /**
   * Rounds towards zero
   */
  DOWN = BigNumber.ROUND_DOWN,
  /**
   * Rounds towards Infinity
   */
  CEIL = BigNumber.ROUND_CEIL,
  /**
   * Rounds towards Infinity
   */
  FLOOR = BigNumber.ROUND_FLOOR,
  /**
   * Rounds towards nearest neighbour.
   * If equidistant, rounds away from zero
   */
  HALF_UP = BigNumber.ROUND_HALF_UP,
  /**
   * Rounds towards nearest neighbour.
   * If equidistant, rounds towards zero
   */
  HALF_DOWN = BigNumber.ROUND_HALF_DOWN,
  /**
   * Rounds towards nearest neighbour.
   * If equidistant, rounds towards even neighbour
   */
  HALF_EVEN = BigNumber.ROUND_HALF_EVEN,
  /**
   * Rounds towards nearest neighbour.
   * If equidistant, rounds towards Infinity
   */
  HALF_CEIL = BigNumber.ROUND_HALF_CEIL,
  /**
   * Rounds towards nearest neighbour.
   * If equidistant, rounds towards Infinity
   */
  HALF_FLOOR = BigNumber.ROUND_HALF_FLOOR,
}

export default ROUNDING;
