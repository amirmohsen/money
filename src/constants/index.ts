/* eslint-disable import/prefer-default-export */
export { default as ROUNDING } from './ROUNDING';
