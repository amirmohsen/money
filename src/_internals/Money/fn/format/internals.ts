import Currency, { CurrencyInputSettings } from '../../../Currency';
import { Money } from '../..';
import { getCurrency } from '../_internals';

/**
 * Get formatting details
 * @param value - Monetary value to be formatted
 * @param settings - Formatting settings, currency code or currency
 * @returns - Formatting details
 * @private
 */
export const getFormattingDetails = (
  value: Money,
  settings?: CurrencyInputSettings|string|Currency,
) => {
  const {
    symbol,
    code,
    thousandsSeparator,
    decimalSeparator,
    decimalDigits,
    pattern,
    formatter,
  } = getCurrency(settings, value).settings;

  let bigNumberValue;

  if (settings === undefined) {
    bigNumberValue = value.amountAsBigNumber;
  } else {
    const BN = value.bigNumberConstructor.clone({
      FORMAT: {
        decimalSeparator,
        groupSeparator: thousandsSeparator,
        groupSize: 3,
        secondaryGroupSize: 0,
        fractionGroupSeparator: ' ',
        fractionGroupSize: 0,
      },
    });

    bigNumberValue = new BN(value.amountAsBigNumber);
  }

  return {
    symbol,
    code,
    decimalDigits,
    thousandsSeparator,
    decimalSeparator,
    pattern,
    formatter,
    value: bigNumberValue,
  };
};
