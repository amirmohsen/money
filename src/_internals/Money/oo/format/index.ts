import { Money } from '../..';
import Currency, { CurrencyInputSettings } from '../../../Currency';
import _format from '../../fn/format';

declare module '../../../Money' {
  interface Money {
    format: (settings?: CurrencyInputSettings | string | Currency) => string;
  }
}

Money.prototype.format = function format(settings?): string {
  return _format(this, settings);
};
