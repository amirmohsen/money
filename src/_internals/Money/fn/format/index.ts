import Currency, { CurrencyInputSettings } from '../../../Currency';
import { Money } from '../..';
import { getFormattingDetails } from './internals';

export interface FormattingSettings {
  thousandsSeparator?: CurrencyInputSettings['thousandsSeparator'];
  decimalSeparator?: CurrencyInputSettings['decimalSeparator'];
  decimalDigits?: CurrencyInputSettings['decimalDigits'];
  pattern?: CurrencyInputSettings['pattern'];
  symbol?: CurrencyInputSettings['symbol'];
  formatter?: CurrencyInputSettings['formatter'];
  parser?: CurrencyInputSettings['parser'];
  code?: CurrencyInputSettings['code'];
}

/**
 * Format money based on settings
 * @param value - Monetary value to be formatted
 * @param settings - Formatting settings, currency code or currency
 * @returns - Formatted money string
 */
const format = (value: Money, settings?: FormattingSettings|string|Currency): string => {
  const {
    symbol,
    code,
    decimalDigits,
    thousandsSeparator,
    decimalSeparator,
    pattern,
    formatter,
    value: innerValue,
  } = getFormattingDetails(value, settings as CurrencyInputSettings);

  const formattedValue = innerValue.absoluteValue().toFormat(decimalDigits);
  const defaultFormatted = pattern
    .replace('%v', formattedValue)
    .replace('%ns', innerValue.isNegative() ? '-' : '')
    .replace('%i', innerValue.isInteger() ? innerValue.toFormat(0) : formattedValue)
    .replace('%s', symbol)
    .replace('%c', code);

  if (typeof formatter === 'function') {
    return formatter({
      value,
      defaultFormatted,
      symbol,
      code,
      decimalDigits,
      thousandsSeparator,
      decimalSeparator,
      pattern,
    });
  }

  return defaultFormatted;
};

export default format;
