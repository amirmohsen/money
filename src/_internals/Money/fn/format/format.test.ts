import { Money } from '../..';
import { USD, GBP, EUR, OMR, JPY } from '../../../constants/ISO_CURRENCIES';
import Currency from '../../../Currency';
import format from '.';

jest.mock('../../../CurrencyStore/internals/getData', () => () => ({
  USD,
  GBP,
  EUR,
  OMR,
  JPY,
}));

describe('format', () => {
  const currencies = [
    'USD',
    'GBP',
    'EUR',
    'OMR',
    'JPY',
  ];

  const values = [
    '100.00',
    '54.42',
    '-120.99',
    '1200145154.42',
  ];

  for (const currency of currencies) {
    describe(`with "${currency}"`, () => {
      for (const value of values) {
        const	money = new Money(value, currency);
        test(`should format ${currency} ${value} correctly`, () => {
          expect(format(money)).toMatchSnapshot();
        });
      }
    });
  }

  describe('when given a currency as settings', () => {

    test('should merge it with the built-in currency to format', () => {
      const customEUR = new Currency({
        ...EUR,
        thousandsSeparator: ',',
        decimalSeparator: '.',
        decimalDigits: 2,
        pattern: '%ns%s%v',
      });
      const money = new Money('45680.90', 'EUR');
      expect(format(money)).toBe('45 680,90 €');
      expect(format(money, customEUR)).toBe('€45,680.90');
    });
  });

  describe('when given a settings object', () => {

    test('should merge it with the built-in currency to format', () => {
      const money = new Money('-45680.90', 'GBP');
      expect(format(money)).toBe('-£45,680.90');
      expect(format(
        money,
        {
          pattern: '%c %ns%v',
        },
      )).toBe('GBP -45,680.90');
    });
  });

  describe('when given a currency code as settings', () => {

    test('should completely override the built-in currency to format', () => {
      const money = new Money('-45680.90', 'GBP');
      expect(format(money)).toBe('-£45,680.90');
      expect(format(money, 'USD')).toBe('-$45,680.90');
    });
  });

  describe('when given a custom formatter', () => {
    const customFormattedValue = 'custom formatting';
    const customFormatter = jest.fn((args: object) => customFormattedValue);
    const customUSD = new Currency({
      ...USD,
      formatter: (args: object) => customFormatter(args),
    });
    const money = new Money('80.90', customUSD);
    const formattedValue = format(money);

    test('should use its return value for formatting', () => {
      expect(formattedValue).toBe(customFormattedValue);
    });

    test('should call it once with the snapshotted arguments', () => {
      expect(customFormatter).toHaveBeenCalledTimes(1);
      expect(customFormatter.mock.calls[0]).toMatchSnapshot();
    });
  });
});
