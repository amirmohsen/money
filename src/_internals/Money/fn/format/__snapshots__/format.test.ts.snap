// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format when given a custom formatter should call it once with the snapshotted arguments 1`] = `
Array [
  Object {
    "code": "USD",
    "decimalDigits": 2,
    "decimalSeparator": ".",
    "defaultFormatted": "$80.90",
    "pattern": "%ns%s%v",
    "symbol": "$",
    "thousandsSeparator": ",",
    "value": Object {
      "amount": "80.90",
      "currency": "USD",
    },
  },
]
`;

exports[`format with "EUR" should format EUR -120.99 correctly 1`] = `"-120,99 €"`;

exports[`format with "EUR" should format EUR 54.42 correctly 1`] = `"54,42 €"`;

exports[`format with "EUR" should format EUR 100.00 correctly 1`] = `"100,00 €"`;

exports[`format with "EUR" should format EUR 1200145154.42 correctly 1`] = `"1 200 145 154,42 €"`;

exports[`format with "GBP" should format GBP -120.99 correctly 1`] = `"-£120.99"`;

exports[`format with "GBP" should format GBP 54.42 correctly 1`] = `"£54.42"`;

exports[`format with "GBP" should format GBP 100.00 correctly 1`] = `"£100.00"`;

exports[`format with "GBP" should format GBP 1200145154.42 correctly 1`] = `"£1,200,145,154.42"`;

exports[`format with "JPY" should format JPY -120.99 correctly 1`] = `"-¥121"`;

exports[`format with "JPY" should format JPY 54.42 correctly 1`] = `"¥54"`;

exports[`format with "JPY" should format JPY 100.00 correctly 1`] = `"¥100"`;

exports[`format with "JPY" should format JPY 1200145154.42 correctly 1`] = `"¥1,200,145,154"`;

exports[`format with "OMR" should format OMR -120.99 correctly 1`] = `"﷼ -120.990"`;

exports[`format with "OMR" should format OMR 54.42 correctly 1`] = `"﷼ 54.420"`;

exports[`format with "OMR" should format OMR 100.00 correctly 1`] = `"﷼ 100.000"`;

exports[`format with "OMR" should format OMR 1200145154.42 correctly 1`] = `"﷼ 1,200,145,154.420"`;

exports[`format with "USD" should format USD -120.99 correctly 1`] = `"-$120.99"`;

exports[`format with "USD" should format USD 54.42 correctly 1`] = `"$54.42"`;

exports[`format with "USD" should format USD 100.00 correctly 1`] = `"$100.00"`;

exports[`format with "USD" should format USD 1200145154.42 correctly 1`] = `"$1,200,145,154.42"`;
